cmd/src/campaign_progress_printer_test.go

```diff
@@ -0,0 +1,249 @@
+package main
+
+import (
+	"runtime"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/sourcegraph/src-cli/internal/campaigns"
+	"github.com/sourcegraph/src-cli/internal/output"
+)
+
+const progressPrinterDiff = `diff --git README.md README.md
+new file mode 100644
+index 0000000..3363c39
+--- /dev/null
++++ README.md
+@@ -0,0 +1,3 @@
++# README
++
++This is the readme
+diff --git a/b/c/c.txt a/b/c/c.txt
+deleted file mode 100644
+index 5da75cf..0000000
+--- a/b/c/c.txt
++++ /dev/null
+@@ -1 +0,0 @@
+-this is c
+diff --git x/x.txt x/x.txt
+index 627c2ae..88f1836 100644
+--- x/x.txt
++++ x/x.txt
+@@ -1 +1 @@
+-this is x
++this is x (or is it?)
+`
+
+func TestCampaignProgressPrinterIntegration(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		t.Skip("Something emits different escape codes on windows.")
+	}
+
+	buf := &ttyBuf{}
+
+	out := output.NewOutput(buf, output.OutputOpts{
+		ForceTTY:    true,
+		ForceColor:  true,
+		ForceHeight: 25,
+		ForceWidth:  80,
+		Verbose:     true,
+	})
+
+	statuses := []*campaigns.TaskStatus{
+		{
+			RepoName:           "github.com/sourcegraph/sourcegraph",
+			StartedAt:          time.Now(),
+			CurrentlyExecuting: "echo Hello World > README.md",
+		},
+		{
+			RepoName:           "github.com/sourcegraph/src-cli",
+			StartedAt:          time.Now().Add(time.Duration(-5) * time.Second),
+			CurrentlyExecuting: "Downloading archive",
+		},
+		{
+			RepoName:           "github.com/sourcegraph/automation-testing",
+			StartedAt:          time.Now().Add(time.Duration(-5) * time.Second),
+			CurrentlyExecuting: "echo Hello World > README.md",
+		},
+	}
+
+	printer := newCampaignProgressPrinter(out, true, 4)
+	printer.forceNoSpinner = true
+
+	// Print with all three tasks running
+	printer.PrintStatuses(statuses)
+	have := buf.Lines()
+	want := []string{
+		"⠋  Executing... (0/3, 0 errored)  ",
+		"│                                                                               ",
+		"├── github.com/sourcegraph/sourcegraph  echo Hello World > README.md          0s",
+		"├── github.com/sourcegraph/src-cli      Downloading archive                   0s",
+		"└── github.com/sourcegraph/automati...  echo Hello World > README.md          0s",
+	}
+	if !cmp.Equal(want, have) {
+		t.Fatalf("wrong output:\n%s", cmp.Diff(want, have))
+	}
+
+	// Now mark the last task as completed
+	statuses[len(statuses)-1] = &campaigns.TaskStatus{
+		RepoName:           "github.com/sourcegraph/automation-testing",
+		StartedAt:          time.Now().Add(time.Duration(-5) * time.Second),
+		FinishedAt:         time.Now().Add(time.Duration(5) * time.Second),
```

Comment:
This could be very subtly flaky if (somehow) the `StartedAt` and `FinishedAt` calls are separated by more than half a second. It might be better to calculate this above this struct as something like:

```go
startedAt := time.Now().Add(time.Duration(-5) * time.Second)
finishedAt := startedAt.Add(time.Duration(10) * time.Second)
```

Of course, I might also just be paranoid.

Reply:
Not paranoid! Good point :)